app/services/uri_service.rb

```diff
@@ -2,25 +2,33 @@ class UriService
   attr_accessor :uri
 
   def initialize(uri)
-    return nil if uri.nil?
+    return nil if uri.blank?
 
     begin 
-      @uri = URI.parse(URI.encode(uri))
+      @uri = URI.parse(uri)
+
+      if @uri.class == URI::Generic
+        # probably missing https. raise error so we can try again
+        raise URI::InvalidURIError
+      end
+
     rescue URI::InvalidURIError
-      nil
+      # Maybe we forgot the scheme... try adding
+      begin
+        @uri = URI.parse("https://" + uri)
+      rescue URI::InvalidURIError
+        # nope!
+        @uri = nil
+        return
+      end
```

Comment:
Let me know if you don't like how this is structured now - I know nested exception handlers aren't 100% ideal. I realized that at the end of the day, all of my fancy regexes were just _approximating_  the `URI.parse` function, so I might as well use it. Basic overview of the new system:

1. Try to parse the URL as provided.
2. If it fails, it's either missing the scheme, or invalid. If the URL has class `URI::Generic` that also tells us there's no scheme, so raise the `InvalidURIError` to run the nested exception handler.
3. Prepend `https://` and try to reparse. If _this_ fails, it's definitely invalid; return.
4. Change the scheme to `https`. If users provide _valid_ HTTP URLs, steps 2 and 3 are skipped, and we just adjust the scheme.